File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.util.*;
import java.io.*;

public class Main{

    public static void main(String[] args) throws IOException {
        Scanner in = new Scanner(System.in);

        boolean isBalanced = true;

        while (true) {
            isBalanced = true;
            String s = in.nextLine();
            Stack<Character> st = new Stack<>();
            if (s.equals(".")) break;
            for (char x : s.toCharArray()) {
                if (x == '(' || x == '[') st.push(x);

                else if (x == ')') {
                    if (st.empty() || st.pop() != '(') {
                        isBalanced = false;
                        break;
                    }
                }

                else if (x == ']') {
                    if (st.empty() || st.pop() != '[') {
                        isBalanced = false;
                        break;
                    }
                }
            }
            if (isBalanced && st.empty()) System.out.println("yes");
            else System.out.println("no");
        }
    }
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <iostream>
#include <bits/stdc++.h>
using namespace std;


int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    while (true){
        stack<char> st;
        string s;
        getline(cin,s);
        bool isValid = true;
        if (s == ".") break;
        for (char tmp : s) {
            if (tmp == '(' || tmp == '[') st.push(tmp);

            else if (tmp == ')'){
                if (st.empty() || st.top() != '('){
                    isValid = false;
                    break;
                }
                st.pop();
            }

            else if (tmp == ']'){
                if (st.empty() || st.top() != '['){
                    isValid = false;
                    break;
                }
                st.pop();
            }
        }
        if (!st.empty()) isValid = false;
        if (isValid) cout << "yes\n";
        else cout << "no\n";
    }
}